### DISCUSSION TITLE
Possible to render a framebuffer inside a div or img?

### DISCUSSION THREAD

**Discussion by ben-metzger-z** [mikke89/RmlUi#632] [Q&A] [answered]
Hi there!

I've been switching my codebase from Dear ImGui to RmlUI for the customisation and more user facing stuff. It's been great so far :) ! Just wondering if there's a way to render from a GL framebuffer into a window? You could do this in ImGui by rendering the framebuffer to a texture and then adding that as an image to ImGui (https://www.codingwiththomas.com/blog/rendering-an-opengl-framebuffer-into-a-dear-imgui-window). Is there something similar I can do with RmlUI? Thanks.

**Comment by mikke89** [comment, marked as answer]
Hey, glad to hear the transition has been smooth so far!

Certainly, there are various ways to achieve this. The easiest I would say is simply to dedicate some kind of texture name for this purpose. Then, in your render interface, whenever you see this texture name, you plug in your framebuffer, and have that one render like other textures. In your document, you would just add this as a normal image or decorator. You might want to give it a name like `/my_framebuffer` so that it is resolved like an absolute path instead of being resolved relative to the document, but those are implementation details.

If you want more power and flexibility, you could implement a [custom decorator](https://mikke89.github.io/RmlUiDoc/pages/cpp_manual/decorators.html). This allows you to customize what you're rendering with your own properties, nicely integrated with style sheets. Also, you can create your framebuffer after layout, when the size is known, if you want it to be reactive to layouting.

Hope that gives you some pointers to start with. Good luck, and let us know if you need some more help implementing this.

**Reply by ben-metzger-z** [reply]
Hello again!

With your pointers I managed to get the first method you suggested working, but its starting to get a little inflexible for my use case. I'm working on getting a custom decorator set up, and I've got that and the DecoratorInstancer working with the RCSS decorator and properties ( very neat :) ).

As far as I can tell, Decorators only have access to `Texture`s rather than a texture handle which could be bound to a framebuffer for rendering. Is this a case where I'll just have to bypass using RmlUI's geometry and rendering methods and render directly with OpenGL?

Thanks

**Reply by mikke89** [reply]
Nice work. And yes, you're exactly on point. Once the RmlUi interface is too limited, you can make your own draw calls to your renderer, directly from the custom decorator.
